File: sahal_pharmacy/static/src/js/pos_pharmacy.js
/** @odoo-module **/

import { patch } from "@web/core/utils/patch";
import { PosStore } from "@point_of_sale/app/services/pos_store";
import { AlertDialog, ConfirmationDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
import { _t } from "@web/core/l10n/translation";

/**
 * Pharmacy rules at the till.
 *
 * This is a convenience layer, NOT the enforcement layer: the browser can always be
 * bypassed, so the authoritative checks live server-side in sale_pos_guard.py, which
 * validates every pos.order on creation. What happens here is telling the cashier
 * early — before payment — instead of letting the sale fail at the end.
 *
 * THIS FILE LOADS INTO EVERY POINT OF SALE ON THE PLATFORM. point_of_sale._assets_pos is
 * one shared bundle: a restaurant, a hotel shop and a hardware counter all download and
 * run this patch on `addLineToOrder`, a core method on the hot path of every sale in
 * every industry.
 *
 * So `pharmacyEnabled()` is the first line of every method here. On a till that is not
 * flagged a pharmacy point of sale, the patch hands straight back to super() and does
 * nothing else — no field reads, no RPCs, no dialogs. A bug in the pharmacy logic can
 * then only break pharmacy tills, which is the whole point.
 */
patch(PosStore.prototype, {
    /** True only on a till configured as a pharmacy point of sale. */
    pharmacyEnabled() {
        return Boolean(this.config?.is_pharmacy);
    },

    /**
     * Intercept adding a product so a restricted medicine cannot be rung up
     * without a verified prescription attached to the order.
     */
    async addLineToOrder(vals, order, opts = {}, configure = true) {
        if (!this.pharmacyEnabled()) {
            return await super.addLineToOrder(...arguments);
        }
        const product = vals.product_id || vals.product_tmpl_id;
        if (product && !(await this.pharmacyCheckProduct(product, order))) {
            return;
        }
        const line = await super.addLineToOrder(...arguments);
        // Warn about near-expiry stock only once the line is actually on the order.
        if (line && product) {
            this.pharmacyWarnExpiry(product);
        }
        return line;
    },

    /** True when the product may be sold on this order. */
    async pharmacyCheckProduct(product, order) {
        if (!this.pharmacyEnabled() || !product.requires_prescription) {
            return true;
        }
        const script = order?.pharmacy_prescription_id;
        if (!script) {
            this.dialog.add(AlertDialog, {
                title: _t("Prescription required"),
                body: _t(
                    "%s is a prescription-only medicine. Attach a verified prescription " +
                        "to this order first (Prescription button), or dispense it from " +
                        "the Pharmacy app.",
                    product.display_name
                ),
            });
            return false;
        }
        if (product.is_controlled) {
            // Controlled drugs are allowed only with an explicit acknowledgement; the
            // server re-checks that the user is a pharmacist regardless.
            return await new Promise((resolve) => {
                this.dialog.add(ConfirmationDialog, {
                    title: _t("Controlled substance"),
                    body: _t(
                        "%s is a controlled substance. It will be written to the " +
                            "controlled drugs register and may only be handed over by a " +
                            "pharmacist. Continue?",
                        product.display_name
                    ),
                    confirm: () => resolve(true),
                    cancel: () => resolve(false),
                });
            });
        }
        return true;
    },

    /** Tell the cashier when the stock behind this product is close to expiring. */
    async pharmacyWarnExpiry(product) {
        if (!this.pharmacyEnabled() || !product.is_medicine) {
            return;
        }
        let warnings = {};
        try {
            warnings = await this.data.call("pharmacy.prescription", "pos_expiry_warnings", [
                [product.id],
            ]);
        } catch {
            return; // never block a sale because the lookup failed
        }
        const batches = warnings[product.id] || [];
        if (!batches.length) {
            return;
        }
        const expired = batches.filter((b) => b.expired);
        const soon = batches.filter((b) => !b.expired);
        const lines = [];
        if (expired.length) {
            lines.push(
                _t("EXPIRED: %s", expired.map((b) => `${b.lot} (${b.expiry})`).join(", "))
            );
        }
        if (soon.length) {
            lines.push(
                _t("Expiring soon: %s", soon.map((b) => `${b.lot} (${b.expiry})`).join(", "))
            );
        }
        this.dialog.add(AlertDialog, {
            title: _t("Check the batch — %s", product.display_name),
            body: lines.join("\n"),
        });
    },

    /** Attach a verified prescription to the current order. */
    async pharmacySelectPrescription() {
        if (!this.pharmacyEnabled()) {
            return;
        }
        const order = this.getOrder();
        if (!order) {
            return;
        }
        const partner = order.getPartner();
        let scripts = [];
        try {
            scripts = await this.data.call(
                "pharmacy.prescription",
                "pos_search_prescriptions",
                [partner ? partner.id : false, null, 20]
            );
        } catch {
            this.dialog.add(AlertDialog, {
                title: _t("Prescriptions"),
                body: _t("Could not load prescriptions. Check the connection and retry."),
            });
            return;
        }
        if (!scripts.length) {
            this.dialog.add(AlertDialog, {
                title: _t("No prescription available"),
                body: partner
                    ? _t(
                          "%s has no verified, unexpired prescription. Create and verify " +
                              "one in the Pharmacy app first.",
                          partner.name
                      )
                    : _t("Select the patient on this order first, then pick their prescription."),
            });
            return;
        }
        // Single match: attach it directly rather than making the cashier pick from
        // a list of one during a queue.
        const chosen = scripts.length === 1 ? scripts[0] : await this.pharmacyPickFrom(scripts);
        if (!chosen) {
            return;
        }
        order.pharmacy_prescription_id = chosen.id;
        if (chosen.allergies) {
            this.dialog.add(AlertDialog, {
                title: _t("Patient allergies — %s", chosen.patient),
                body: chosen.allergies,
            });
        }
    },

    async pharmacyPickFrom(scripts) {
        const list = scripts
            .map((s, i) => `${i + 1}. ${s.name} — ${s.patient}`)
            .join("\n");
        return await new Promise((resolve) => {
            this.dialog.add(ConfirmationDialog, {
                title: _t("Select a prescription"),
                body: _t("Verified prescriptions for this customer:\n%s\n\nThe most recent will be used.", list),
                confirm: () => resolve(scripts[0]),
                cancel: () => resolve(null),
            });
        });
    },
});

File: sahal_pharmacy/models/pos_pharmacy.py
# -*- coding: utf-8 -*-
"""Point of sale integration (BRD 5.5, PRD 35-45).

The till already does barcodes, payments, split payments, discounts, returns and
receipts natively. What it does not know is which products are restricted, so this
module ships the pharmacy attributes down to the client and gives the cashier a way
to look up a verified prescription without leaving the POS.

ONE POS SERVES EVERY INDUSTRY ON THIS PLATFORM — a restaurant, a hotel shop, a hardware
counter and a pharmacy all run the same point_of_sale module and share its asset bundle.
So everything here obeys one rule:

    Nothing pharmaceutical reaches a till that is not flagged `is_pharmacy`.

That applies to the fields loaded, the models loaded, and the JavaScript: the client
patch returns immediately on a non-pharmacy till, before it touches anything. It matters
because this module has already broken every POS on the platform once — an unconditional
`_load_pos_data_models` override made pharmacy.prescription load into every session and
crashed the loader (PROD_PENDING_CHANGES section 1).

The one thing NOT gated on the flag is compliance enforcement. Whether a prescription-only
medicine may be sold is a legal question about the PRODUCT, not a preference of the till,
so sale_pos_guard.py checks every order that actually contains restricted products —
otherwise unticking a checkbox would be a bypass. Orders with no pharmacy products exit
that check immediately.
"""

from odoo import models, fields, api, _

# Batches expiring within this window make the till warn the cashier.
POS_EXPIRY_WARN_DAYS = 60


def _is_pharmacy_config(env, config):
    """True only for a till explicitly flagged as a pharmacy point of sale.

    The POS loader passes a recordset to some hooks and a bare id to others
    (product.template gets `config_id`), so both are accepted rather than assumed.
    """
    if not config:
        return False
    if isinstance(config, int):
        config = env['pos.config'].browse(config)
    return bool(config.exists() and config.is_pharmacy)


class ProductTemplate(models.Model):
    _inherit = 'product.template'

    @api.model
    def _load_pos_data_fields(self, config_id):
        """Ship the pharmacy attributes — to pharmacy tills only.

        A restaurant till has no use for `controlled_schedule`, and loading it there is
        both waste on every product record and pharmacy schema leaking into industries
        that never asked for it.
        """
        fields_list = super()._load_pos_data_fields(config_id)
        if not _is_pharmacy_config(self.env, config_id):
            return fields_list
        extra = [
            'is_medicine', 'pharma_type', 'requires_prescription', 'is_controlled',
            'pharma_generic_name', 'brand_name', 'requires_cold_storage',
            'strength',
        ]
        return list(fields_list) + [f for f in extra if f not in fields_list]


class ProductProduct(models.Model):
    _inherit = 'product.product'

    @api.model
    def _load_pos_data_fields(self, config):
        fields_list = super()._load_pos_data_fields(config)
        if not _is_pharmacy_config(self.env, config):
            return fields_list
        extra = ['is_medicine', 'pharma_type', 'requires_prescription', 'is_controlled',
                 'pharma_generic_name', 'brand_name']
        return list(fields_list) + [f for f in extra if f not in fields_list]


class PosConfig(models.Model):
    _inherit = 'pos.config'

    @api.model
    def _load_pos_data_fields(self, config):
        """The client needs the flag itself, or the JS cannot tell which till it is on."""
        return super()._load_pos_data_fields(config) + ['is_pharmacy']

    is_pharmacy = fields.Boolean(
        string='Pharmacy Point of Sale',
        help="Turn on the pharmacy features for this till: restricted-item flags, "
             "prescription lookup, and expiry warnings. Leave off for ordinary retail "
             "or restaurant tills so they don't load pharmacy data.")


class PosSession(models.Model):
    _inherit = 'pos.session'

    @api.model
    def _load_pos_data_models(self, config):
        """Make pharmacy.prescription available, but only to pharmacy tills.

        `config` is the pos.config recordset. Two conditions must both hold:
        - the till is explicitly flagged a pharmacy point of sale, and
        - the user actually has pharmacy access.

        The group check keeps prescriptions out of a session whose user has no
        pharmacy role at all — even if an
        is_pharmacy flag lingered on a config. Loading this model into every POS —
        retail, restaurant, etc. — is both wasteful and (before this guard) fatal.
        """
        data = super()._load_pos_data_models(config)
        if config.is_pharmacy and self.env.user.has_group('sahal_pharmacy.group_pharmacy_cashier'):
            data = data + ['pharmacy.prescription']
        return data


class PharmacyPrescriptionPos(models.Model):
    # pos.load.mixin gives the model _load_pos_data_search_read / _read, which the POS
    # loader calls on every model in _load_pos_data_models. Without it the till aborts
    # with "'pharmacy.prescription' object has no attribute '_load_pos_data_search_read'".
    _name = 'pharmacy.prescription'
    _inherit = ['pharmacy.prescription', 'pos.load.mixin']

    @api.model
    def _load_pos_data_domain(self, data, config):
        """Only scripts that may actually be dispensed today reach the till."""
        return [
            ('state', 'in', ('verified', 'partially_dispensed')),
            '|', ('valid_until', '=', False),
                 ('valid_until', '>=', fields.Date.context_today(self)),
        ]

    @api.model
    def _load_pos_data_fields(self, config):
        return ['id', 'name', 'patient_id', 'doctor_id', 'state',
                'valid_until', 'has_controlled']

    @api.model
    def pos_search_prescriptions(self, partner_id=None, query=None, limit=20):
        """Look up dispensable prescriptions from the POS.

        A separate RPC rather than relying only on preloaded records: a busy pharmacy
        writes prescriptions all day, and a cashier must be able to find one created
        after the session was opened.
        """
        domain = [
            ('state', 'in', ('verified', 'partially_dispensed')),
            '|', ('valid_until', '=', False),
                 ('valid_until', '>=', fields.Date.context_today(self)),
        ]
        if partner_id:
            domain.append(('patient_id', '=', int(partner_id)))
        if query:
            domain += ['|', ('name', 'ilike', query), ('patient_id.name', 'ilike', query)]
        scripts = self.search(domain, limit=int(limit), order='prescription_date desc')
        return [{
            'id': rx.id,
            'name': rx.name,
            'patient': rx.patient_id.display_name,
            'doctor': rx.doctor_id.display_name,
            'valid_until': rx.valid_until and str(rx.valid_until) or '',
            'has_controlled': rx.has_controlled,
            'allergies': rx.patient_id.allergies or '',
            'lines': [{
                'product_id': line.product_id.id,
                'product': line.product_id.display_name,
                'remaining': line.qty_remaining,
            } for line in rx.line_ids if line.qty_remaining > 0],
        } for rx in scripts]

    @api.model
    def pos_expiry_warnings(self, product_ids):
        """Batches of these products expiring soon or already expired.

        Returned to the till so the cashier is told BEFORE taking payment, rather than
        discovering it when the delivery is validated.
        """
        if not product_ids:
            return {}
        horizon = fields.Datetime.add(fields.Datetime.now(), days=POS_EXPIRY_WARN_DAYS)
        quants = self.env['stock.quant'].sudo().search([
            ('product_id', 'in', [int(p) for p in product_ids]),
            ('quantity', '>', 0),
            ('location_id.usage', '=', 'internal'),
            ('lot_id.expiration_date', '!=', False),
            ('lot_id.expiration_date', '<=', horizon),
        ])
        now = fields.Datetime.now()
        result = {}
        for quant in quants:
            entry = result.setdefault(quant.product_id.id, [])
            entry.append({
                'lot': quant.lot_id.name,
                'expiry': str(fields.Date.to_date(quant.lot_id.expiration_date)),
                'quantity': quant.quantity,
                'expired': quant.lot_id.expiration_date < now,
            })
        return result


class PosOrderLine(models.Model):
    _inherit = 'pos.order.line'

    @api.model
    def _load_pos_data_fields(self, config):
        fields_list = super()._load_pos_data_fields(config)
        if not _is_pharmacy_config(self.env, config):
            return fields_list
        return list(fields_list) + ['pharmacy_prescription_line_id']

    pharmacy_prescription_line_id = fields.Many2one(
        'pharmacy.prescription.line', string='Prescription Line', index=True,
        help='Prescription line this sale fulfils, when sold against a script.')
